database/sqlite/sqlite_functions.c

```diff
@@ -0,0 +1,1224 @@
+    if (unlikely(rc != SQLITE_OK))
+        error_report("Error %d while closing the SQLite database", rc);
+    return;
+}
+
+/*
+ * Return the database size in MiB
+ */
+int sql_database_size(void)
+{
+    sqlite3_stmt *chk_size;
+    int rc;
+
+    rc = sqlite3_prepare_v2(db_meta, "pragma page_count;", -1, &chk_size, 0);
+    if (rc != SQLITE_OK)
+        return 0;
+
+    if (sqlite3_step(chk_size) == SQLITE_ROW)
+        page_count = sqlite3_column_int(chk_size, 0);
+
+    sqlite3_finalize(chk_size);
+
+    rc = sqlite3_prepare_v2(db_meta, "pragma freelist_count;", -1, &chk_size, 0);
+    if (rc != SQLITE_OK)
+        return 0;
+
+    if (sqlite3_step(chk_size) == SQLITE_ROW)
+        free_page_count = sqlite3_column_int(chk_size, 0);
+
+    sqlite3_finalize(chk_size);
+
+    if (unlikely(!page_size)) {
+        rc = sqlite3_prepare_v2(db_meta, "pragma page_size;", -1, &chk_size, 0);
+        if (rc != SQLITE_OK)
+            return 0;
+
+        if (sqlite3_step(chk_size) == SQLITE_ROW)
+            page_size = (uint32_t)sqlite3_column_int(chk_size, 0);
+
+        sqlite3_finalize(chk_size);
+        desired_pages = (sqlite_disk_quota_mb * 0.95) * (1024 * 1024 / page_size);
+        info(
+            "Database desired size is %u pages (page size is %u bytes). Current size is %u pages (includes %u free pages)",
+            desired_pages, page_size, page_count, free_page_count);
+    }
+
+    return ((page_count - free_page_count) / 1024) * (page_size / 1024);
+}
+
+/*
+ * Add a UUID in memory for fast lookup
+ *
+ * Caller should lock the uuid_cache
+ *   Under host holds charts
+ *   Under chart holds dimensions
+ *
+ */
+
+static inline void add_in_uuid_cache(struct uuid_cache **uuid_cache, uuid_t *uuid, const char *type, const char *id, const char *name)
+{
+    //info("add_in_uuid_cache: Adding type=%s, id=%s, name=%s", type?type:"Dimension", id, name?name:"<null>");
+    struct uuid_cache *item = mallocz(sizeof(*item));
+    uuid_copy(item->uuid, *uuid);
+    item->id = id ? strdupz(id) : NULL;
+    item->type = type ? strdupz(type) : NULL;
+    item->name = name ? strdupz(name) : NULL;
+    item->next = *uuid_cache;
+    *uuid_cache = item;
+    return;
+}
+
+/*
+ * Destroy the uuid cache used during startup
+ * Normally it should be empty if all charts and dimensions are active
+ * TODO: Cleanup the cache after the agent starts to save memory
+ * charts / dimensions that will be activated will just query the database
+ */
+void free_uuid_cache(struct uuid_cache **uuid_cache)
+{
+    struct uuid_cache *item;
+    while ((item = *uuid_cache)) {
+        *uuid_cache = item->next;
+        freez(item->id);
+        freez(item->type);
+        freez(item->name);
+        freez(item);
+    }
+    return;
+}
+
+/*
+ * Find a type,id,name in cache -- remove if found
+ * Caller must lock the uuid_cache
+ */
+uuid_t *find_in_uuid_cache(struct uuid_cache **uuid_cache, const char *type, const char *id, const char *name)
```

Comment:
Should we optimize with sth. like skip list?

Reply:
I don't see it as an urgent issue because that list is used only during init of charts / dimensions and it progressively shrinks as items are found. 

However because charts and dimensions are created in a more or less consistent order (when a collector starts) I will add a small change to the query itself which will result in the items stored in reverse order in the list.

Doing a quick test on an agent with metadata log to migrate, the item search was in the first entry of the list
- Before the query change --> 14% of the time
- After the query change --> 74% of the time

That jumps to 92% for an agent that starts fresh (dbengine / metadata log files deleted)